test/jdk/java/util/zip/ZipFile/Zip64SizeTest.java

```diff
@@ -20,117 +20,160 @@
-     * @throws IOException
+     * Validate that if the 'uncompressed size' of a ZIP CEN header is 0xFFFFFFFF, then the
+     * actual size is retrieved from the corresponding ZIP64 Extended information field.
+     *
+     * @throws IOException if an unexpected IOException occurs
      */
     @Test
-    private static void validateZipEntrySizes() throws IOException {
-        createFiles();
+    public void validateZipEntrySizes() throws IOException {
         createZipFile();
         System.out.println("Validating Zip Entry Sizes");
-        try (ZipFile zip = new ZipFile(ZIP_FILE_NAME)) {
-            ZipEntry ze = zip.getEntry(LARGE_FILE_NAME);
+        try (ZipFile zip = new ZipFile(ZIP_FILE.toFile())) {
+            ZipEntry ze = zip.getEntry("first");
             System.out.printf("Entry: %s, size= %s%n", ze.getName(), ze.getSize());
-            assertTrue(ze.getSize() == LARGE_FILE_SIZE);
-            ze = zip.getEntry(SMALL_FILE_NAME);
+            assertEquals(CONTENT.length, ze.getSize());
+            ze = zip.getEntry("second");
             System.out.printf("Entry: %s, size= %s%n", ze.getName(), ze.getSize());
-            assertTrue(ze.getSize() == SMALL_FILE_SIZE);
-
+            assertEquals(CONTENT.length, ze.getSize());
         }
     }
 
     /**
-     * Delete the files created for use by the test
-     * @throws IOException if an error occurs deleting the files
+     * Create ZIP file with a CEN entry where the 'uncompressed size' is stored
+     * in the ZIP64 field. This makes the ZIP64 data block 8 bytes long which
+     * triggers the regression described in 8226530.
+     *
+     * The CEN entry for the "first" entry will have the following structure:
+     * (Note the CEN 'Uncompressed Length' being 0xFFFFFFFF and the ZIP64
+     * 'Uncompressed Size' being 5)
+     *
+     * 0081 CENTRAL HEADER #1     02014B50
+     * 0085 Created Zip Spec      14 '2.0'
+     * 0086 Created OS            00 'MS-DOS'
+     * [...]
+     * 0091 CRC                   F7D18982
+     * 0095 Compressed Length     00000007
+     * 0099 Uncompressed Length   FFFFFFFF
+     * [..]
+     * 00AF Filename              'first'
+     * 00B4 Extra ID #0001        0001 'ZIP64'
+     * 00B6   Length              0008
+     * 00B8   Uncompressed Size   0000000000000005
+     *
+     * @throws IOException if an error occurs creating the ZIP File
      */
-    private static void deleteFiles() throws IOException {
-        Files.deleteIfExists(Path.of(ZIP_FILE_NAME));
-        Files.deleteIfExists(Path.of(LARGE_FILE_NAME));
-        Files.deleteIfExists(Path.of(SMALL_FILE_NAME));
+    private static void createZipFile() throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try (ZipOutputStream zos = new ZipOutputStream(baos)) {
+
+            ZipEntry e1 = new ZipEntry("first");
+            // Make room for an 8-byte ZIP64 extra field
+            e1.setExtra(createOpaqueExtra((short) Long.BYTES));
```

Comment:
> Hello Eirik, I couldn't understand why we first add a opaque extra field first and then update it to be a zip64 extra field. Why do we do this?

`ZipEntry.setExtra` processes the byte array argument, looking for Zip64 extended fields which it can extract the size fields from. To prevent this parsing from happening, we temporarily use the `unknown` tag.

In this particular case, `ZipExtra.setExtra` actually ends up skipping this processing (because `isLOC == true` and it has a guard for the block size being `>= 16`).

However, I prefer the test to not depend too much on the details of `setExtra` Zip64 processing. This trick is used in other tests as well and may be copied over to a test where the conditions are not the same.

I have refactored a bit and added some code comments to help explain the use of the 'unknown' tag.

Do you think this makes sense?